hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/DeletedBlockLogStateManagerImpl.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+import org.apache.hadoop.hdds.utils.db.BatchOperationHandler;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.hdds.utils.db.TypedTable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Proxy;
+
+import static org.apache.hadoop.hdds.scm.ScmConfigKeys.OZONE_SCM_BLOCK_DELETION_MAX_RETRY;
+import static org.apache.hadoop.hdds.scm.ScmConfigKeys.OZONE_SCM_BLOCK_DELETION_MAX_RETRY_DEFAULT;
+
+public class DeletedBlockLogStateManagerImpl
+    implements DeletedBlockLogStateManager {
+
+  public static final Logger LOG =
+      LoggerFactory.getLogger(DeletedBlockLogStateManagerImpl.class);
+
+  private final Table<Long, DeletedBlocksTransaction> deletedTable;
+  private final BatchOperationHandler batchHandler;
+  private final int maxRetry;
+
+  public DeletedBlockLogStateManagerImpl(
+      ConfigurationSource conf,
+      Table<Long, DeletedBlocksTransaction> deletedTable,
+      BatchOperationHandler batchHandler) {
+    this.maxRetry = conf.getInt(OZONE_SCM_BLOCK_DELETION_MAX_RETRY,
+        OZONE_SCM_BLOCK_DELETION_MAX_RETRY_DEFAULT);
+    this.deletedTable = deletedTable;
+    this.batchHandler = batchHandler;
+  }
+
+  public TableIterator<Long, TypedTable.KeyValue<Long,
+      DeletedBlocksTransaction>> getReadOnlyIterator() {
+    return new TableIterator<Long, TypedTable.KeyValue<Long,
+        DeletedBlocksTransaction>>() {
+      private TableIterator<Long,
+          ? extends Table.KeyValue<Long, DeletedBlocksTransaction>> iter =
+          deletedTable.iterator();
+
+      @Override
+      public boolean hasNext() {
+        return iter.hasNext();
+      }
+
+      @Override
+      public TypedTable.KeyValue<Long, DeletedBlocksTransaction> next() {
+        return iter.next();
+      }
+
+      @Override
+      public void close() throws IOException {
+        iter.close();
+      }
+
+      @Override
+      public void seekToFirst() {
+        iter.seekToFirst();
+      }
+
+      @Override
+      public void seekToLast() {
+        iter.seekToLast();
+      }
+
+      @Override
+      public TypedTable.KeyValue<Long, DeletedBlocksTransaction> seek(
+          Long key) throws IOException {
+        return iter.seek(key);
+      }
+
+      @Override
+      public Long key() throws IOException {
+        return iter.key();
+      }
+
+      @Override
+      public TypedTable.KeyValue<Long, DeletedBlocksTransaction> value() {
+        return iter.value();
+      }
+
+      @Override
+      public void removeFromDB() throws IOException {
+        throw new UnsupportedOperationException("read-only");
+      }
+    };
+  }
+
+  @Override
+  public void addTransactionsToDB(DeleteBlocksCommandProto txs)
+      throws IOException {
+    BatchOperation batch = batchHandler.initBatchOperation();
```

Comment:
You need try-with-resource together with the `BatchOperation`, otherwise the rocksdb might crash.

See
```
  public void addTransactions(Map<Long, List<Long>> containerBlocksMap)
      throws IOException {
    lock.lock();
    try {
      try(BatchOperation batch =
          scmMetadataStore.getStore().initBatchOperation()) {
        for (Map.Entry< Long, List< Long > > entry :
            containerBlocksMap.entrySet()) {
          long nextTXID = scmMetadataStore.getNextDeleteBlockTXID();
          DeletedBlocksTransaction tx = constructNewTransaction(nextTXID,
              entry.getKey(), entry.getValue());
          scmMetadataStore.getDeletedBlocksTXTable().putWithBatch(batch,
              nextTXID, tx);
        }
        scmMetadataStore.getStore().commitBatchOperation(batch);
      }
    } finally {
      lock.unlock();
    }
  }
```